File: scripts/discord_news_bot.py
# ruff: noqa: I001
import os
import subprocess
from pathlib import Path
from shutil import which
from typing import Any

import discord
from discord import app_commands

import sys  # noqa: I001

ROOT = Path(__file__).resolve().parents[1]
sys.path.append(str(ROOT))

from src.config import load_config  # noqa: E402


def load_environment(path: Path) -> None:
    if not path.exists():
        return
    for line in path.read_text(encoding="utf-8").splitlines():
        if "=" not in line or line.lstrip().startswith("#"):
            continue
        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip())


def workflow_command(settings: dict[str, Any], query: str) -> list[str]:
    command = list(settings.get("workflow_command", ["uv", "run", "python", "-m", "src.main"]))
    env_key = settings.get("workflow_command_env", "UV_BIN")
    override = os.getenv(env_key)
    if override:
        command[0] = override
    else:
        resolved = which(command[0])
        if resolved:
            command[0] = resolved
    flag = settings.get("workflow_argument_flag", "--query")
    command.extend([flag, query])
    return command


def render(template: str, query: str) -> str:
    return template.replace("{query}", query)


def register_news_command(tree: app_commands.CommandTree, settings: dict[str, Any], project_root: Path) -> None:
    response_template = settings.get("response_template", "Generating video for: {query}")
    starter_template = settings.get("starter_template", "Starting generation for: {query}")
    thread_prefix = settings.get("thread_prefix", "video-")
    thread_name_limit = int(settings.get("thread_name_limit", 50))
    thread_message = settings.get("thread_message", "Progress will be posted here.")
    command = settings.get("command_name", "news")
    description = settings.get("command_description", "Generate news video")

    @tree.command(name=command, description=description)
    async def handle(interaction: discord.Interaction, query: str) -> None:
        await interaction.response.defer(ephemeral=True)
        starter = await interaction.channel.send(render(starter_template, query))
        thread_name = f"{thread_prefix}{query}"[:thread_name_limit]
        thread = await starter.create_thread(name=thread_name)
        await thread.send(render(thread_message, query))
        subprocess.Popen(workflow_command(settings, query), cwd=str(project_root))
        await interaction.followup.send(render(response_template, query), ephemeral=True)


def create_client(settings: dict[str, Any], project_root: Path) -> discord.Client:
    intents = discord.Intents.default()
    intents.message_content = True

    class Client(discord.Client):
        def __init__(self) -> None:
            super().__init__(intents=intents)
            self.tree = app_commands.CommandTree(self)

        async def setup_hook(self) -> None:
            register_news_command(self.tree, settings, project_root)
            await self.tree.sync()

    return Client()


def main() -> None:
    config = load_config()
    news_settings = config.get("news_bot", {})
    if not news_settings:
        return

    project_root = Path.cwd()
    env_file = news_settings.get("environment_file", "config/.env")
    load_environment(Path(env_file))

    token = os.environ.get(news_settings.get("token_env", "DISCORD_TOKEN"))
    if not token:
        return

    client = create_client(news_settings, project_root)
    client.run(token)


if __name__ == "__main__":
    main()

File: scripts/discord_news_bot.ts
import "dotenv/config";
import { spawn } from "node:child_process";
import {
	Client,
	GatewayIntentBits,
	REST,
	Routes,
	SlashCommandBuilder,
	type TextChannel,
} from "discord.js";
import { loadConfig } from "../src/core.js";

async function main() {
	const config = loadConfig();
	const settings = config.news_bot || {};
	if (!settings) {
		console.log("No news_bot settings found.");
		return;
	}

	const token = process.env.DISCORD_TOKEN || process.env[settings.token_env];
	if (!token) {
		console.error("No DISCORD_TOKEN found in env.");
		return;
	}

	const client = new Client({ intents: [GatewayIntentBits.Guilds] });

	const commandName = settings.command_name || "news";
	const description = settings.command_description || "Generate news video";
	const strTemplate =
		settings.starter_template || "Starting generation for: {query}";
	const resTemplate =
		settings.response_template || "Generating video for: {query}";
	const threadPrefix = settings.thread_prefix || "video-";
	const threadMsg = settings.thread_message || "Progress will be posted here.";

	const commands = [
		new SlashCommandBuilder()
			.setName(commandName.toLowerCase())
			.setDescription(description)
			.addStringOption((option) =>
				option
					.setName("query")
					.setDescription("Topic to generate news about")
					.setRequired(true),
			),
	];

	const rest = new REST({ version: "10" }).setToken(token);

	client.once("ready", async () => {
		console.log(`Logged in as ${client.user?.tag}!`);
		console.log("Refreshing application (/) commands.");
		if (client.user) {
			await rest.put(Routes.applicationCommands(client.user.id), {
				body: commands,
			});
		}
	});

	client.on("interactionCreate", async (interaction) => {
		if (!interaction.isChatInputCommand()) return;

		if (interaction.commandName === commandName) {
			const query = interaction.options.getString("query")!;

			await interaction.deferReply({ ephemeral: true });

			const channel = interaction.channel as TextChannel;
			if (!channel) {
				await interaction.editReply("Can only be used in a text channel.");
				return;
			}

			const starterMsg = await channel.send(
				strTemplate.replace("{query}", query),
			);

			const threadName = `${threadPrefix}${query}`.substring(0, 50);
			const thread = await starterMsg.startThread({
				name: threadName,
				autoArchiveDuration: 60,
			});

			await thread.send(threadMsg.replace("{query}", query));

			console.log(`Spawning workflow for: ${query}`);
			const child = spawn("npx", ["tsx", "src/index.ts", query], {
				cwd: process.cwd(),
				stdio: "ignore",
				detached: true,
			});
			child.unref();

			await interaction.editReply(resTemplate.replace("{query}", query));
		}
	});

	client.login(token);
}

main().catch(console.error);
